state-chain/pallets/cf-account-roles/src/lib.rs

```diff
@@ -152,6 +186,84 @@ pub mod pallet {
 			Self::deposit_event(Event::VanityNameSet { account_id, name });
 			Ok(())
 		}
+
+		/// Derives a sub-account by the given origin account id and a sub-account index.
+		/// Stores the account id against the sub-account index and origin account id.
+		///
+		/// The sub account is unmutual associated with the origin/parent account and the creation
+		/// fails if the sub-account already exists for the given index. It's possible to derive 256
+		/// sub-accounts per parent account. The parent account can execute calls on behalf of the
+		/// sub-account (see as_sub_account).
+		///
+		/// Events:
+		/// - `SubAccountCreated` if the sub-account is created.
+		///
+		/// Errors:
+		/// - `SubAccountAlreadyExists` if the sub-account already exists for the given index.
+		/// - `SubAccountIdDerivationFailed` if the sub-account id derivation fails.
+		#[pallet::call_index(1)]
+		#[pallet::weight(T::WeightInfo::derive_sub_account())]
+		pub fn derive_sub_account(
+			origin: OriginFor<T>,
+			sub_account_index: SubAccountIndex,
+		) -> DispatchResult {
+			let account_id = ensure_signed(origin)?;
+			ensure!(
+				!SubAccounts::<T>::contains_key(&account_id, sub_account_index),
+				Error::<T>::SubAccountAlreadyExists
+			);
+			let sub_account_id: T::AccountId = Decode::decode(&mut TrailingZeroInput::new(
+				(*b"chainflip/subaccount", account_id.clone(), sub_account_index)
+					.blake2_256()
+					.as_ref(),
+			))
+			.map_err(|_| Error::<T>::SubAccountIdDerivationFailed)?;
+			SubAccounts::<T>::insert(&account_id, sub_account_index, &sub_account_id);
+			Self::deposit_event(Event::SubAccountCreated {
+				account_id: account_id.clone(),
+				sub_account_id,
+				sub_account_index,
+			});
+			Ok(())
+		}
+
+		/// Executes a call on behalf of a sub-account. The account is getting identified by the
+		/// passed sub-account index.
+		///
+		/// The call is executed with the sub-account's account id as the dispatch origin.
+		///
+		/// Events:
+		/// - `SubAccountCallExecuted` if the call is executed.
+		///
+		/// Errors:
+		/// - `UnknownAccount` if the sub-account is not found.
+		/// - `FailedToExecuteCallOnBehalfOfSubAccount` if the call fails.
+		#[pallet::call_index(2)]
+		#[pallet::weight(T::WeightInfo::as_sub_account().saturating_add(call.get_dispatch_info().weight))]
+		pub fn as_sub_account(
+			origin: OriginFor<T>,
+			sub_account_index: SubAccountIndex,
+			call: Box<<T as Config>::RuntimeCall>,
+		) -> DispatchResult {
+			let mut origin = origin;
+			let account_id = ensure_signed(origin.clone())?;
+			let sub_account_id = SubAccounts::<T>::get(&account_id, sub_account_index);
+			ensure!(sub_account_id.is_some(), Error::<T>::UnknownAccount);
+			let sub_account_id = sub_account_id.unwrap();
+			origin.set_caller_from(frame_system::RawOrigin::Signed(sub_account_id.clone()));
+
+			call.clone()
+				.dispatch(origin)
+				.map_err(|_| Error::<T>::FailedToExecuteCallOnBehalfOfSubAccount)?;
```

Comment:
```suggestion
			call.clone()
				.dispatch(origin)?;
```

I feel like it would be useful to know the real error instead of swallowing it.